### DISCUSSION TITLE
How to resolve different cache keys in list of array?

### DISCUSSION THREAD

**Discussion by nghiepdev** [urql-graphql/urql#1420] [Q&A]
Hi guys,

I have an example:

 **GraphQL schema**
```gql
type Product {
   id: ID!
   name: String!
   price: Int!
}

type Order {
   id: ID!
   total: Int!
   products: [Product]
}

type Query {
   orders: [Order]
}
```

**Request list of orders query**

```gql
const GET_ORDER_LIST = gql`
  query GetOrderList{
    orders {
      id
      total
      products {
         id
         name
         price
      }
    }
  }
`;
```

**Sample data**
```
Order 1: {
 id: 1,
 product_ids: [1, 2]
}

Order 2: {
 id: 2,
 product_ids: [2, 3]
}
```

**Problem:**  `Order 1` and `Order 2` both have the `product_id=2` but different `price` value.

When rendering to the frontend the `price` of `product_id=2` in the of`Order2` overwrite the `price` of `Order1`.
Because the `Product` type has the same id value.
How to resolve different the `Product` type in the list of orders?

**Comment by nghiepdev** [comment]
Currently, I resolve the `Product` type is `null` instead of `id`.
```js
const cache = cacheExchange({
   keys: {
      Product(data) {
         return null;
         // return data.id;
      }
   }
});
```

**Reply by kitten** [reply]
I think this is a change you'll want to make to your schema. All your types have a field called `id` which indicates that it's a globally unique value, however it's instead an index.

That also means that you can't actually share these types in a normalised form. They're dependent on the Liston which they are if that's their only ID field, so returning null here is correct which makes them embedded data

**Reply by nghiepdev** [reply]
tks @kitten 
But it would be nice if there was another way like that
```js
const cache = cacheExchange({
   keys: {
      Product(data) {
         if(isChildOfAnother) {
             return null;
         }
         return data.id;
      }
   }
});
```

**Reply by kitten** [reply]
Oh, I see; while this is in theory possible for us to add, we do have some opinions on how our normalised caching is implemented which is why it's not possible and those are all based on lowering the likelihood of mistakes and predictability.

I'd actually say this case sounds like you may want to introduce a new type that only is _like_ products or add some kind of `ProductEdge` which is common in Relay-compliant schemas, since that'd give you a type that can carry information about the specific item in the list

The reason why I think it's odd to have this choice in the first place is that as a human reading this I see two identical products with different data, so a solution like the above (or even a separate edge or line item) result also intuitively makes more sense
